spdx-utils/src/main/kotlin/model/SpdxDocument.kt

```diff
@@ -138,5 +139,20 @@ data class SpdxDocument(
         require(dataLicense.isNotBlank()) { "The data license must not be blank." }
 
         require(documentNamespace.isNotBlank()) { "The document namespace must not be blank." }
+
+        val duplicatePackages = packages.getDuplicates { it.spdxId }
```

Comment:
I roughly remember a thought from a while ago which was questioning whether it makes sense
to throw runtime exceptions for any kind of consistency violation.

My outcome was that it makes sense to have both:

1. checks which result in a failure parsing the document
2. checks which result in a failed validity check (This would require a separate component which tells you the issues with a document, e.g. like a linter).

Do you think above makes sense and if so, under which of above two categories do you believe these added checks would best fall under?

Reply:
To me, 1. would be syntax checks that are already implicitly performed by Jackson upon deserialization. Point 2. would then be (semantical) consistency / sanity checks, and my changes would fall into that category.

I added my changes to the constructor because similar checks were already present. But I'd also be fine with moving all those checks to a `validate()` function that throws an exception with a respective message, and leave it to the user of this class to call (or not) the `validate()` function. Would you prefer that?

Reply:
hm, I find it quite hard where to draw the line between these two categories.
What Jackson catches also depends on how detailed we represent the model.For example an Spdx identifier is represented as String while it could be modeled as separate class, in particular decomposing into a (valid) prefix bit and a suffix. Anyhow, as you named it, syntax checks, would probably be a good term for category 1, or maybe also well formedness. 

From a use case perspective I'd say category 1 should be checks which fit in any thinkable use case while category 2 would not fit in any. 

Anyhow, I'd need to take time to think closer about these two categories and what kind of checks we have.
The checks you add could maybe named `referential integrity` as `duplicates` break referencing themselves. I'd be fine with putting the checks here for now and leaving this decision up for later when it becomes clearer.

Reply:
> From a use case perspective I'd say category 1 should be checks which fit in any thinkable use case while category 2 would not fit in any.

I don't believe that's a good distinction, when you say that not every thinkable use case must pass the checks from category 2. I believe every thinkable use case should also pass the checks from category 2.

Basically, to me it boils down to distinguishing between

1. things that you *can* put into the format (but that do not necessarily make sense, like homepages that do not have a valid URL format or duplicate package IDs), and
2. things where the content actually makes sense, is usable, and consistent.

> I'd be fine with putting the checks here for now and leaving this decision up for later when it becomes clearer.

Agreed.